memtable/memtable_garbage_test.cc

```diff
@@ -0,0 +1,282 @@
+  //                                                      insertion type"
+  //             + VarintLength(val_size)           --> = min # of bytes to
+  //             store val_size
+  //             + val_size                         --> = actual value string
+  // For example, in our situation, "key1" : size 4, "value1" : size 6
+  // (the terminating null characters are not copied over to the memtable).
+  // And therefore encoded_len = 1 + (4+8) + 1 + 6 = 20 bytes per entry.
+  // However in terms of raw data contained in the memtable, and written
+  // over to the SSTable, we only count internal_key_size and val_size,
+  // because this is the only raw chunk of bytes that contains everything
+  // necessary to reconstruct a user entry: sequence number, insertion type,
+  // key, and value.
+
+  // To test the relevance of our Memtable garbage statistics,
+  // namely MEMTABLE_DATA_BYTES and MEMTABLE_GARBAGE_BYTES,
+  // we insert 3 distinct K-V pairs NUM_REPEAT times.
+  // I chose NUM_REPEAT=20,000 such that no automatic flush is
+  // triggered (the number of bytes in the memtable is therefore
+  // well below any meaningful heuristic for a memtable of size 64MB).
+  // As a result, since each K-V pair is inserted as a payload
+  // of 18 meaningful bytes (sequence number, insertion type,
+  // key, and value = (8) + 4 + 6 = 18), MEMTABLE_DATA_BYTES
+  // should be equal to 20,000 * 3 * 18 = 1,080,000 bytes
+  // and MEMTABLE_GARBAGE_BYTES = MEMTABLE_DATA_BYTES - (3*18)
+  //                            = 1,079,946 bytes.
+
+  const size_t NUM_REPEAT = 20000;
+  const char KEY1[] = "key1";
+  const char KEY2[] = "key2";
+  const char KEY3[] = "key3";
+  const char VALUE1[] = "value1";
+  const char VALUE2[] = "value2";
+  const char VALUE3[] = "value3";
+  const uint64_t USEFUL_PAYLOAD_BYTES =
+      strlen(KEY1) + strlen(VALUE1) + strlen(KEY2) + strlen(VALUE2) +
+      strlen(KEY3) + strlen(VALUE3) + 3 * sizeof(uint64_t);
+  const uint64_t EXPECTED_MEMTABLE_DATA_BYTES =
+      NUM_REPEAT * USEFUL_PAYLOAD_BYTES;
+  const uint64_t EXPECTED_MEMTABLE_GARBAGE_BYTES =
+      (NUM_REPEAT - 1) * USEFUL_PAYLOAD_BYTES;
+
+  // Insertion of of K-V pairs, multiple times.
+  for (size_t i = 0; i < NUM_REPEAT; i++) {
+    ASSERT_OK(db_->Put(WriteOptions(), KEY1, VALUE1));
+    ASSERT_OK(db_->Put(WriteOptions(), KEY2, VALUE2));
+    ASSERT_OK(db_->Put(WriteOptions(), KEY3, VALUE3));
+  }
+  // We assert that the K-V pairs have been successfully inserted.
+  std::string value;
+  ASSERT_OK(db_->Get(ReadOptions(), KEY1, &value));
+  ASSERT_EQ(value, VALUE1);
+  ASSERT_OK(db_->Get(ReadOptions(), KEY2, &value));
+  ASSERT_EQ(value, VALUE2);
+  ASSERT_OK(db_->Get(ReadOptions(), KEY3, &value));
+  ASSERT_EQ(value, VALUE3);
+
+  // Force flush to SST. Increments the statistics counter.
+  if (db_ != nullptr) {
+    FlushOptions flush_opt;
+    flush_opt.wait = true;  // function returns once the flush is over.
```

Comment:
This is default and we make that assumption all over the place, so Flush() (from DBTestBase) should be fine.